src/compiler/types/lines_2000_4000.rs

```diff
@@ -534,6 +534,84 @@ impl TypeAliasDeclaration {
     }
 }
 
+pub type CommentKind = SyntaxKind; /*SyntaxKind.SingleLineCommentTrivia | SyntaxKind.MultiLineCommentTrivia*/
+
+pub struct CommentRange {
+    pos: Cell<isize>,
+    end: Cell<isize>,
+    has_trailing_new_line: Option<bool>,
+    kind: CommentKind,
+}
+
+impl CommentRange {
+    pub fn new(
+        kind: CommentKind,
+        pos: isize,
+        end: isize,
+        has_trailing_new_line: Option<bool>,
+    ) -> Self {
+        Self {
+            kind,
+            pos: Cell::new(pos),
+            end: Cell::new(end),
+            has_trailing_new_line,
+        }
+    }
+}
+
+impl TextRange for CommentRange {
+    fn pos(&self) -> isize {
+        self.pos.get()
+    }
+
+    fn set_pos(&self, pos: isize) {
+        self.pos.set(pos);
+    }
+
+    fn end(&self) -> isize {
+        self.end.get()
+    }
+
+    fn set_end(&self, end: isize) {
+        self.end.set(end);
+    }
+}
+
+pub type SourceTextAsChars = Vec<char>;
+
+pub fn text_len(text: &SourceTextAsChars) -> usize {
+    text.len()
+}
+
+pub fn maybe_text_char_at_index(text: &SourceTextAsChars, index: usize) -> Option<char> {
+    text.get(index).map(|ch| *ch)
+}
+
+pub fn text_char_at_index(text: &SourceTextAsChars, index: usize) -> char {
+    maybe_text_char_at_index(text, index).unwrap()
+}
+
+pub fn text_substring(text: &SourceTextAsChars, start: usize, end: usize) -> String {
+    text[start..end].into_iter().collect()
+}
+
+pub fn text_str_num_chars(text: &str, start: usize, end: usize) -> usize {
+    text[start..end].chars().count()
+}
+
+pub trait SourceFileLike {
+    fn text(&self) -> &str;
+    fn text_as_chars(&self) -> &SourceTextAsChars;
+    fn maybe_line_map(&self) -> RefMut<Option<Vec<usize>>>;
+    fn line_map(&self) -> Ref<Vec<usize>>;
+    fn maybe_get_position_of_line_and_character(
```

Comment:
In the Typescript version `SourceFileLike` has an optional `getPositionOfLineAndCharacter()` method, I guess the most literal translation of that would be for `maybe_get_position_of_line_and_character` to be an `Option` of a closure type (that you can then call if it's present) but seems easier to have it always be implemented and just always return `None` if it's effectively not implemented